Include full register state in ARM Cortex M Exception Stack Frame (ESF)

chrisc11 opened the pull request:
Draft PR for changes proposed in https://github.com/zephyrproject-rtos/zephyr/issues/27495

carlescufi commented:
@chrisc11 thanks for the PR. Please follow the [guidelines](https://docs.zephyrproject.org/latest/contribute/index.html#commit-guidelines) for the commit message formatting (thanks for the extensive doc in the commit message by the way).

I think this is mature enough to be a non-draft PR by the way.

andrewboie commented on arch/arm/core/aarch32/fatal.c:
```diff
@@ -18,24 +18,25 @@ LOG_MODULE_DECLARE(os);
-		esf->basic.a4, esf->basic.ip, esf->basic.lr);
-	LOG_ERR(" xpsr:  0x%08x", esf->basic.xpsr);
+		hw_esf->basic.a4, hw_esf->basic.ip, hw_esf->basic.lr);
+	LOG_ERR(" xpsr:  0x%08x", hw_esf->basic.xpsr);
```
Since we now have these extra registers available we should dump them here along with the ones we already print out.

tyrantlizard77 reviewed (commented):
I think this needs to be a configuration option, because if I understand this correctly, we're pushing extra data that we don't strictly need to handle the exception. Not all exceptions are fatal and produce a crash.

This is a problem common to all architectures. Nios II has CONFIG_EXTRA_EXCEPTION_INFO to toggle this. x86 32-bit seems to just push the extra registers unconditionally. x86 64-bit hangs this on CONFIG_EXCEPTION_DEBUG. This seems to be done ad hoc per arch currently.

I'd sugget up-leveling CONFIG_EXTRA_EXCEPTION_INFO to arch/Kconfig and conditionally build this against that. That way if this isn't needed for debugging we won't have extra overhead for exception processing. Later we can go in and change other arches to all use CONFIG_EXTRA_EXCEPTION_INFO

chrisc11 commented:
Thanks for the feedback and info @andrewboie!

> I think this needs to be a configuration option, because if I understand this correctly, we're pushing extra data that we don't strictly need to handle the exception. Not all exceptions are fatal and produce a crash.

You are right that depending on other configurations options the system may recover from the fault instead of rebooting. In these scenarios it could still be useful to collect all the registers to enable recovery of a backtrace of what caused the non-fatal fault. I do restore the register state in fault_s.S to deal with the scenario where one returns from the fault.

I can definitely put this behind a `CONFIG_EXTRA_EXCEPTION_INFO` as you suggest though so folks trying to optimize the processing time of faults can ensure the collection is disabled.

ceolin commented on arch/arm/core/aarch32/cortex_m/fault_s.S:
```diff
@@ -76,14 +76,50 @@ SECTION_SUBSEC_FUNC(TEXT,__fault,z_arm_debug_monitor)
+	mov r1, r9
+	mov r0, r8
+	push {r0-r3, lr}
+	push {r3-r7}
```
shouldn't be `push {r4-r7}` ?

chrisc11 replied:
good observation @ceolin  ... I tried to add a comment on line 101 about this. Per the [ARM procedure call standard ](https://documentation-service.arm.com/static/5f201281bb903e39c84d7eae), "The stack must be double-word aligned" at a public interface, so I pushed one extra register to satisfy the requirement for the call to `z_arm_fault`

ceolin commented on include/arch/arm/aarch32/exc.h:
```diff
@@ -91,7 +92,26 @@ struct __esf {
+
+struct __full_esf {
+	struct __esf *exception_frame;
+	bool nested_exc;
```
I'd let the padding in the end of the struct

andrewboie commented:
> In these scenarios it could still be useful to collect all the registers to enable recovery of a backtrace of what caused the non-fatal fault.

What I'm getting at is that in some cases it wouldn't be needed at all and would be undesirable, like a recoverable MPU fault (`arch_user_string_nlen` fixups), or a PendSV exception to trigger a context switch (we do this weird exception dance on ARM Cortex-M for context switching)

> I can definitely put this behind a CONFIG_EXTRA_EXCEPTION_INFO as you suggest though so folks trying to optimize the processing time of faults can ensure the collection is disabled.

Sweet thanks!

chrisc11 commented:
> > In these scenarios it could still be useful to collect all the registers to enable recovery of a backtrace of what caused the non-fatal fault.
> 
> What I'm getting at is that in some cases it wouldn't be needed at all and would be undesirable, like a recoverable MPU fault (`arch_user_string_nlen` fixups)

Makes sense, the `arch_user_string_nlen` check using the MPU is nifty!

> or a PendSV exception to trigger a context switch (we do this weird exception dance on ARM Cortex-M for context switching)

hrm, I see context switching handling via the [z_arm_pendsv handler here](https://github.com/chrisc11/zephyr/blob/master/arch/arm/core/aarch32/swap_helper.S#L52) / PendSV exception but I don't see a context switch path that would use the fault handler in fault_s.S. Do you happen to know what context switch path would use the fault handler from fault_s.S?

andrewboie commented:
> hrm, I see context switching handling via the [z_arm_pendsv handler here](https://github.com/chrisc11/zephyr/blob/master/arch/arm/core/aarch32/swap_helper.S?rgh-link-date=2020-08-11T22%3A37%3A30Z#L52) / PendSV exception but I don't see a context switch path that would use the fault handler in fault_s.S. Do you happen to know what context switch path would use the fault handler from fault_s.S?

I might be mistaken on this bit, @ioannisg knows this stuff better than I. If the way these exceptions are used for context switching doesn't use this path then that's really good.

ioannisg commented:
Thanks for the contribution @chrisc11 , I am going to take a look as soon as possible.

In the mean time you could give it a try to fix CI issues, and, in particular, ensure that the tests/arch/arm family of tests, and also kernel/fatal execute successfully.

chrisc11 commented:
> Thanks for the contribution @chrisc11 , I am going to take a look as soon as possible.

Awesome, thanks @ioannisg!

> In the mean time you could give it a try to fix CI issues, and, in particular, ensure that the tests/arch/arm family of tests, and also kernel/fatal execute successfully.

 Definitely, will do!

ioannisg commented:
Some generic input - I 'll do a proper pass later

- I don't think the option should be `default y`. And it should not be specific for ARM (this is actually problematic, since ARM covers AARC64, and AARCH32 Cortex-A,R as well, while the support is added in Cortex-M only)
I expect many developers want to have fault-dump but they are OK with the current information. So the extra info can be `default n,` 
- I would remove exc_return from the structure of __callee_saved register (R14 is a callee-saved, strictly speaking, but exc_return is a special argument, and it is _already_ passed to the fault handler today as an argument). I would like to continue having exc_return as a separate argument so the fault.S and have the pointer to the callee_saved structure as the 4th argument, #ifdef EXTRA_INFO

chrisc11 commented:
> Some generic input - I 'll do a proper pass later

Thanks for the thoughts @ioannisg !

>   I don't think the option should be `default y`.

Okay, I switched it to off by default (and added a test config to test the path where it is set to `y`)

>   I would like to continue having exc_return as a separate argument as it is now in fault.S, and have the pointer to the callee_saved structure as the 4th argument, #ifdef EXTRA_INFO

Sounds reasonable to me. If it is okay with you, I would like to include the exc_return as part of `z_arch_esf_t` when EXTRA_INFO is enabled. The stacking information encoded in exc_return is needed to determine the `$sp` value prior to exception entry and that info is needed in order to recover the backtrace leading up to the fault.

I'll push a commit with these two changes so you can take a look!

ioannisg commented on arch/arm/core/aarch32/cortex_m/fault_s.S:
```diff
@@ -76,6 +76,53 @@ SECTION_SUBSEC_FUNC(TEXT,__fault,z_arm_debug_monitor)
 #endif /* CONFIG_ARMV6_M_ARMV8_M_BASELINE */
 SECTION_SUBSEC_FUNC(TEXT,__fault,z_arm_exc_spurious)
 
+#if defined(CONFIG_EXTRA_EXCEPTION_INFO)
```
I would be really fond of a single code flow, with in-lining the instructions needed for CONFIG_EXTRA_EXCEPTION_INFO

What you gain is
- you do not need to have #else code, i.e. !CONFIG_EXTRA_EXCEPTION_INFO, so we make it clear that the option only adds extra code.
- you do not need to recover the MSP in the Baseline code path

Something along the following:

```
    mrs  r0, MSP
    mrs  r1, PSP
    push {r0, lr}
#if defined(CONFIG_EXTRA_EXCEPTION_INFO)
#if defined(CONFIG_ARMV6_M_ARMV8_M_BASELINE)
    mov  r3, r11
    mov  r2, r10
    push {r2, r3}
    mov  r3, r9
    mov  r2, r8
    push {r2, r3}
    push {r4-r7}
#elif defined(CONFIG_ARMV7_M_ARMV8_M_MAINLINE)
    push {r4-r11}
#endif
    mrs  r3, MSP /* pointer to the dumped callee-saved registers in main stack */
#endif /* CONFIG_EXTRA_EXCEPTION_INFO */
    mov r2, lr /* EXC_RETURN */
    bl z_arm_fault
#if defined(CONFIG_EXTRA_EXCEPTION_INFO)
    /* Restore dumped called saved registers */
#if defined(CONFIG_ARMV6_M_ARMV8_M_BASELINE)
    pop { r4-r7 }
    pop { r0-r3 }
    mov r11, r3
    mov r10, r2
    mov r9, r1
    mov r8, r0
#elif defined(CONFIG_ARMV7_M_ARMV8_M_MAINLINE)
    pop {r4-r11}
#endif
#endif /* CONFIG_EXTRA_EXCEPTION_INFO */
```

ioannisg replied:
Actually, @chrisc11 I am wondering whether we need to "restore" the callee-saved registers, here. Are they not preserved by the ARM procedure calling standard conventions? Could we just pop them, to restore the stack pointer?

ioannisg commented on arch/arm/core/aarch32/cortex_m/fault.c:
```diff
@@ -929,52 +932,48 @@ static inline z_arch_esf_t *get_esf(uint32_t msp, uint32_t psp, uint32_t exc_ret
  * @param psp PSP value immediately after the exception occurred
  * @param exc_return EXC_RETURN value present in LR after exception entry.
- *
+ * @param callee_regs Registers that are not part of the hardware esf (r4-r11)
```
Just make this explicit:
`@param callee_regs Callee-saved registers (R4-R11)`

ioannisg commented on arch/arm/core/aarch32/cortex_m/fault.c:
```diff
@@ -929,52 +932,48 @@ static inline z_arch_esf_t *get_esf(uint32_t msp, uint32_t psp, uint32_t exc_ret
+ * @param callee_regs Registers that are not part of the hardware esf (r4-r11)
  */
-void z_arm_fault(uint32_t msp, uint32_t psp, uint32_t exc_return)
+void z_arm_fault(uint32_t msp, uint32_t psp, uint32_t exc_return
```
Instead, could you keep the same signature in all cases and then just "not use" the last argument #if !defined(CONFIG_EXTRA_EXCEPTION_INFO)??

ioannisg commented on arch/arm/core/aarch32/cortex_m/fault.c:
```diff
@@ -929,52 +932,48 @@ static inline z_arch_esf_t *get_esf(uint32_t msp, uint32_t psp, uint32_t exc_ret
 	 * the fault handling process.
 	 */
-	z_arch_esf_t esf_copy;
+	z_arch_esf_t esf;
```
Why are you renaming this? It is still a stacked copy, so why not make it clear with the name?

ioannisg commented on include/arch/arm/aarch32/exc.h:
```diff
@@ -91,7 +92,28 @@ struct __esf {
+	uint32_t r11;
+};
+
+struct __full_esf {
```
I am not sure this is the right thing to do.
The callee-saved registers are not part of the the exception stack frame, and adding the container here is confusing.

In my opinion, the best thing is to keep the callee-saved registers information outside the exception stack frame.
The values of the registers are _absolutely_ not needed by the kernel to determine the root cause of the errors; the values should not at all be used in fault.c

That said, I would propose:
- use the existing callee-saved register structure, defined for Cortex-m:
```
struct _callee_saved {
	uint32_t v1;  /* r4 */
	uint32_t v2;  /* r5 */
	uint32_t v3;  /* r6 */
	uint32_t v4;  /* r7 */
	uint32_t v5;  /* r8 */
	uint32_t v6;  /* r9 */
	uint32_t v7;  /* r10 */
	uint32_t v8;  /* r11 */
	uint32_t psp; /* r13 */
};

typedef struct _callee_saved _callee_saved_t;
```

- In fault.S store and pass a pointer to such a struct in z_arm_fault
- In fault.c do nothing, but _transparently_ pass this information in fatal.c

chrisc11 replied:
Thanks for the thoughts @ioannisg! One goal I'm trying to solve for here is passing the information into the weak function `k_sys_fatal_error_handler`. This allows an end user to implement a handler that saves off the the stack & register state needed to recover a full backtrace after a crash has taken place. If the callee saved state is removed from `z_arch_esf_t`, how would you propose passing that state so an end user can collect it?

ioannisg replied:
Ah, I see, so it is not sufficient for the fatal.c to process the information, right? I didn't think about it.
In that case, I guess we need to include the callee-saved regs into the z_arch_esf_t structure. Check if this can be done by adding a struct _callee_saved pointer so we can reuse the already defined structure.

Please, investigate, if this can be done in a clean and transparent way, that is
- fault.S to include a pointer to a structure that holds the callee-saved regs as we discussed above, when calling z_arm_fault
- fault.c to _copy_ the callee-saved regs into the stacked copy of z_arch_esf_t right before calling z_arm_fatal_error (ideally, no more changes should be needed)

Could this work?

ioannisg reviewed (changes requested):
@chrisc11 I had a deeper look into this draft PR
I acknowledge the value of passing the callee-saved register dump into the fault handling mechanism.

I have concerns on _how_ this should be done, see my review comments.
Since the callee-saved regs are not part of the exception stack frame, I think they should not be added in the corresponding structure.

I am leaning towards a solution that is less intrusive; generate the callee-saved struct in fault.S, by-pass fault.c and use the information in fatal.c as desired.

How does this sound to you?

ioannisg commented on include/arch/arm/aarch32/exc.h:
```diff
@@ -73,25 +73,45 @@ GTEXT(z_arm_exc_exit);
+	uint32_t xpsr;
+};
+
 struct __esf {
```
Do we really need to keep __esf and __extended_esf?

could we try to use only the "extended" esf where we clearly document that the extra information is at the end?

Otherwise I am OK with this re-factoring.

ioannisg commented on include/arch/arm/aarch32/exc.h:
```diff
@@ -73,25 +73,45 @@ GTEXT(z_arm_exc_exit);
 	uint32_t undefined;
 #endif
+#if defined(CONFIG_EXTRA_EXCEPTION_INFO)
+	/* Additional register state that is not stacked by
```
Can you put all this to a struct, to make it easier to read?
E.g. `struct _extra_info` or `struct extra_exception_info`

ioannisg commented on include/arch/arm/aarch32/exc.h:
```diff
@@ -73,25 +73,45 @@ GTEXT(z_arm_exc_exit);
 };
 
-typedef struct __esf z_arch_esf_t;
+typedef struct __extended_esf z_arch_esf_t;
```
So if my proposal is feasible (pls, investigate), this could be left as it is now. :)

ioannisg commented on arch/arm/core/aarch32/cortex_m/fault.c:
```diff
@@ -959,7 +961,13 @@ void z_arm_fault(uint32_t msp, uint32_t psp, uint32_t exc_return)
+	esf_copy.callee = callee_regs;
+	esf_copy.exc_return = exc_return;
+	esf_copy.msp = msp;
+#endif
```
Can you add an #else and set _callee_regs to UNUSED argument?

ioannisg commented on arch/arm/core/aarch32/cortex_m/fault.c:
```diff
@@ -131,7 +131,7 @@ LOG_MODULE_DECLARE(os);
 
 #if (CONFIG_FAULT_DUMP == 1)
-static void fault_show(const z_arch_esf_t *esf, int fault)
+static void fault_show(const struct __esf *esf, int fault)
```
I believe (actually, I hope :) ) that all these argument signature changes won't be needed if we accept the convention that the z_arch_esf_t will always be __esf, and contain the additional info at the end.

chrisc11 replied:
> I think I like the new approach, but I have one last request for investigation, which I think will make the patch look even more "light".
> that is, try to define the extended register info as a last field inside the _esf (document this properly) and solely use the z_arch_esf_t in fault.c.

thanks for the review @ioannisg, yes I think that would work. The only nuance I see with that approach is the extra data would only be valid in the `esf_copy` passed in z_arm_fatal_error. That is, the pointer returned from `get_esf` would not contain the info. If just documenting that and reusing the struct sounds better to you, happy to make the change.

> Should we schedule this enhancement for Zephyr v2.4.0 release? Merge window closes in about 10 days, so if you have the bandwidth to work on this patch, we should be able to have it for next release.

Sounds great, I'll aim to wrap up the final changes you are suggesting over the next day or so!

ioannisg reviewed (commented):
Thanks for the re-work @chrisc11 
I think I like the new approach, but I have one last request for investigation, which I think will make the patch look even more "light".
that is, try to define the extended register info as a last field inside the _esf (document this properly) and solely use the z_arch_esf_t  in fault.c.

And please, move the PR to non-draft, since this is really in-review, not work-in-progress.

Should we schedule this enhancement for Zephyr v2.4.0 release? Merge window closes in about 10 days, so if you have the bandwidth to work on this patch, we should be able to have it for next release.

chrisc11 commented:
@ioannisg  this should be ready for a re-review.

I've rebased the pr, updated to an approach such that `__extended_esf` could be removed, and updated the PR so it's no longer marked as a draft.

ioannisg commented on arch/arm/core/aarch32/cortex_m/fault.c:
```diff
@@ -963,7 +965,20 @@ void z_arm_fault(uint32_t msp, uint32_t psp, uint32_t exc_return)
+		.exc_return = exc_return,
+		.msp = msp
+	};
+#endif
```
Pls, add a comment on the endif,
/* CONFIG_EXTRA_EXCEPTION_INFO */

ioannisg reviewed (commented):
Thanks for the updates, @chrisc11 , i think this looks very good. I have some minor suggestions, see in the relevant comments. Besides that, I have also:
 - one suggestion: maybe in fatal.c you can dump the extra exception info as well, so we can visualize the extra information, before the call to z_fatal_error().
- one request: I'd like you to actually test the new feature, not just provide a test-case that builds with CONFIG_EXTRA_EXCEPTION_INFO. For example, you could write a test case that sets the R4-R11 to some predefined values, forces a fault (e.g. with null-pointer exception, invalid SVC call, BusFault, etc.) and the handler reads out and verifies the passed callee-saved registers with the predefined structure. You can add this in any relevant test suite under tests/arch/arm

chrisc11 commented:
> one suggestion: maybe in fatal.c you can dump the extra exception info as well, so we can visualize the extra information, before the call to z_fatal_error().

sounds good, done!
 
>  one request: I'd like you to actually test the new feature, not just provide a test-case that builds with CONFIG_EXTRA_EXCEPTION_INFO. [...] You can add this in any relevant test suite under tests/arch/arm

Good idea @ioannisg ! I added a new test case to arm_interrupt.c for this. I added checks to make sure the extra info is populated correctly as well as the basic stack frame when CONFIG_EXTRA_EXCEPTION_INFO=n since I didn't see a test case for that yet either.

ioannisg reviewed (commented):
Looks good, thanks @chrisc11 
I 'll do a proper pass on the PR tomorrow morning

ioannisg commented on tests/arch/arm/arm_interrupt/src/arm_interrupt.c:
```diff
@@ -22,13 +98,90 @@ void k_sys_fatal_error_handler(unsigned int reason, const z_arch_esf_t *pEsf)
+	);
+}
+
+void test_arm_esf_collection(void)
```
Maybe you could run the test only when CONFIG_EXTRA_EXCEPTION_INFO is defined, otherwise you can skip the test. 
(see how other tests in this suite are "compile-time" skipped)

I suggest this cause I don't see much value on running the test if this Kconfig macro is not defined.

chrisc11 replied:
hmm, my thought here was this let's us test that the copy of the esf was done correctly for the `CONFIG_EXTRA_EXCEPTION_INFO=n` codepath. Otherwise, there's no check in place that the basic stack frame is populated correctly for that configuration.

ioannisg replied:
Alright, no strong opinion here, you can leave that :)

ioannisg reviewed (approved):
Thanks for the work @chrisc11 , looks good to me.

You need to rework your commit message formats; right now you've got 3 commits with the same title, different bodies. There are some guidelines for commit messages (e.g. title line up to 70 characters, body lines up to 65, I think).

You could either submit all in single commit, or have multiple commits for different file patches, up to you.

chrisc11 committed:
arch: arm: Collect full register state in Cortex-M Exception Stack Frame

To debug hard-to-reproduce faults/panics, it's helpful to get the full
register state at the time a fault occurred. This enables recovering
full backtraces and the state of local variables at the time of a
crash.

This PR introduces a new Kconfig option, CONFIG_EXTRA_EXCEPTION_INFO,
to facilitate this use case. The option enables the capturing of the
callee-saved register state (r4-r11 & exc_return) during a fault. The
info is forwarded to `k_sys_fatal_error_handler` in the z_arch_esf_t
parameter. From there, the data can be saved for post-mortem analysis.

To test the functionality a new unit test was added to
tests/arch/arm_interrupt which verifies the register contents passed
in the argument match the state leading up to a crash.

Signed-off-by: Chris Coleman <chris@memfault.com>

chrisc11 commented:
> Thanks for the work @chrisc11 , looks good to me.

Thanks for all the feedback @ioannisg !

> You need to rework your commit message formats; right now you've got 3 commits with the same title, different bodies. 

Yeah, I wasn't sure what the usual flow during code reviews was for that.

> There are some guidelines for commit messages (e.g. title line up to 70 characters, body lines up to 65, I think).

I see there are some commit message styles [enforced in CI](https://github.com/zephyrproject-rtos/zephyr/blob/master/scripts/gitlint/zephyr_commit_rules.py) so I was using those length limits as my guide

> You could either submit all in single commit, or have multiple commits for different file patches, up to you.

Sounds great, I'm personally a fan of the [one-idea-is-one-commit](https://secure.phabricator.com/book/phabflavor/article/recommendations_on_revision_control/) style workflow anyway so I've pushed an update that folds the change into a single commit!

andrewboie commented on tests/arch/arm/arm_interrupt/testcase.yaml:
```diff
@@ -11,3 +11,12 @@ tests:
+    tags: arm interrupt ignore_faults
+    arch_allow: arm
+    extra_configs:
+      - CONFIG_NO_OPTIMIZATIONS=y
```
why do we need to disable optimizations to test this?

chrisc11 replied:
no reason, just copied from the config above. I'll remove all the extra config settings except `CONFIG_EXTRA_EXCEPTION_INFO=y` and push an update

MaureenHelm reviewed (approved):
neat, thanks @chrisc11!

nashif commented:
@chrisc11  nice, this https://github.com/zephyrproject-rtos/zephyr/pull/27589 is related I guess..

stephanosio reviewed (approved):
LGTM

ioannisg commented:
> @chrisc11 nice, this #27589 is related I guess..

@nashif indeed, it is related, although this PR focuses on adding extra information to the fault handler, while #27589 puts the standard ESF info to a core dump mechanism. I guess the core-dump can be extended to include the extra information, (since now we will have the framework for that...)

chrisc11 commented:
> > @chrisc11 nice, this #27589 is related I guess..
> 
> @nashif indeed, it is related, although this PR focuses on adding extra information to the fault handler, while #27589 puts the standard ESF info to a core dump mechanism. I guess the core-dump can be extended to include the extra information, (since now we will have the framework for that...)

Neat! Yes, the changes look very complementary. With the changes maybe in this PR, the missing registers in [aarch32/cortex_m/coredump.c](https://github.com/zephyrproject-rtos/zephyr/blob/181d073/arch/arm/core/aarch32/cortex_m/coredump.c#L49-L55) could now be populated instead of getting stubbed out.
